pkg/standalone-utils/test/VaultActionsStableJoinExit.test.ts

```diff
@@ -0,0 +1,757 @@
+  setChainedReferenceContents,
+  toChainedReference,
+} from './helpers/chainedReferences';
+import TypesConverter from '@balancer-labs/v2-helpers/src/models/types/TypesConverter';
+import { Account } from '@balancer-labs/v2-helpers/src/models/types/types';
+import {
+  setupRelayerEnvironment,
+  encodeJoinPool,
+  encodeExitPool,
+  PoolKind,
+  approveVaultForRelayer,
+  getJoinExitAmounts,
+  encodeSwap,
+} from './VaultActionsRelayer.setup';
+
+describe('stable pools', () => {
+  let vault: Vault;
+  let tokens: TokenList;
+  let relayer: Contract, relayerLibrary: Contract;
+  let user: SignerWithAddress, other: SignerWithAddress;
+  let sender: Account, recipient: Account;
+  let poolIdStable: string;
+  let poolIdStable2: string;
+  let bptIndex: number;
+  let bptIndex2: number;
+
+  before('setup environment', async () => {
+    ({ user, other, vault, relayer, relayerLibrary } = await setupRelayerEnvironment());
+  });
+
+  before('setup common recipient', () => {
+    // All the tests use the same recipient; this is a simple abstraction to improve readability.
+    recipient = user;
+  });
+
+  sharedBeforeEach('set up pools', async () => {
+    tokens = (await TokenList.create(['DAI', 'CDAI'])).sort();
+    await tokens.mint({ to: user, amount: fp(1000) });
+    await tokens.approve({ to: vault, from: user });
+
+    const stablePool = await StablePool.create({
+      tokens,
+      vault,
+    });
+
+    bptIndex = await stablePool.getBptIndex();
+    const equalBalances = Array.from({ length: tokens.length + 1 }).map((_, i) => (i == bptIndex ? 0 : fp(1000)));
+    await stablePool.init({ recipient: user.address, initialBalances: equalBalances, from: user });
+
+    poolIdStable = await stablePool.getPoolId();
+    // Create a second one with the same tokens, for chaining
+    const stablePool2 = await StablePool.create({
+      tokens,
+      vault,
+    });
+
+    bptIndex2 = await stablePool2.getBptIndex();
+    const equalBalances2 = Array.from({ length: tokens.length + 1 }).map((_, i) => (i == bptIndex2 ? 0 : fp(1000)));
+    await stablePool2.init({ recipient: user.address, initialBalances: equalBalances2, from: user });
+
+    poolIdStable2 = await stablePool2.getPoolId();
+  });
+
+  describe('join pool', () => {
+    const amountInDAI = fp(2);
+    const amountInCDAI = fp(5);
+
+    context('when caller is not authorized', () => {
+      it('reverts', async () => {
+        await expect(
+          relayer.connect(other).multicall([
+            await encodeJoinPool(vault, relayerLibrary, {
+              poolKind: PoolKind.COMPOSABLE_STABLE_V2,
+              poolId: poolIdStable,
+              userData: '0x',
+              sender: user.address,
+              recipient,
+            }),
+          ])
+        ).to.be.revertedWith('Incorrect sender');
+      });
+    });
+
+    context('when caller is authorized', () => {
+      context('sender = user', () => {
+        beforeEach(async () => {
+          sender = user;
```

Comment:
```suggestion
      context('sender = user', () => {
        let sender: SignerWithAddress;
      
        beforeEach(async () => {
          sender = user;
```

You'll need to repeat this in other tests, but it'll make to so we're not sharing the `sender` variable accross blocks, which can lead to user error and hard to debug scenarios (if e.g. we forget to initialize it in one block, since some tests would then depend on unrelated tests, based on execution order)